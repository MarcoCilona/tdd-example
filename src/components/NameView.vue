<template>
  <div>
    <p data-test-id="paragraphTest">Hello {{ name }}</p>
    <button data-test-id='greetingsButton'
      @click='handleMoreGreetingsClick'>More greeting please!</button>
  </div>
</template>

<script lang="ts">
import {
  Component, Prop, Vue,
} from 'vue-property-decorator';

@Component
export default class NameView extends Vue {
  @Prop() name!: string;

  handleMoreGreetingsClick() {
    this.$router.push({
      path: `/more-greetings/${this.name}`,
    });
  }
}
</script>
