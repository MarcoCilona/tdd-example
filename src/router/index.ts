import Vue from 'vue';
import VueRouter, { RouteConfig } from 'vue-router';

import Home from '@/views/Home.vue';
import MoreGreetings from '@/views/MoreGreetings.vue';

Vue.use(VueRouter);

export const routes: Array<RouteConfig> = [
  {
    path: '/',
    component: Home,
  },
  {
    path: '/more-greetings/:name',
    name: 'MoreGreetings',
    component: MoreGreetings,
  },
];

const router = new VueRouter({
  routes,
});

export default router;
