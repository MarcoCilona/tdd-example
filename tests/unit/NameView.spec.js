import { createLocalVue, shallowMount } from '@vue/test-utils';

// Components
import NameView from '@/components/NameView.vue';

import VueRouter from 'vue-router';
import { routes } from '@/router/index';

const localVue = createLocalVue();
localVue.use(VueRouter);

/**
 * 1. Components does exists
 * 2. Component contains a paragraph element
 * 3. Component has a name property
 * 4. Component is printing 'Hello ${name}' inside a paragraph
 * 5. Component contains button to go to more greetings
 * 5. When button is clicked the app gets redirect
 * 6. The new route contains the name property
 */
describe('NewView tests', () => {
  let wrapper;
  let router;
  beforeEach(() => {
    router = new VueRouter({ routes, mode: 'abstract' });
    wrapper = shallowMount(NameView, {
      localVue,
      router,
      propsData: {
        name: '',
      },
    });
  });

  it('Components does exists', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Components contains a paragraph element', () => {
    expect(wrapper.find('[data-test-id="paragraphTest"]').exists()).toBe(true);
  });

  it('Components has a name property', () => {
    expect(wrapper.props().name).not.toBeUndefined();
  });

  it('Name prop is shown inside paragraph', async () => {
    const paragraph = wrapper.find('[data-test-id="paragraphTest"]').element;
    const nameToBeShown = 'Digitiamo';
    wrapper.setProps({
      name: nameToBeShown,
    });

    await wrapper.vm.$nextTick();
    expect(paragraph.textContent).toEqual('Hello Digitiamo');
  });

  it('Component contains button to go to more greetings', () => {
    expect(wrapper.find('[data-test-id="greetingsButton"]').exists()).toBe(true);
  });

  it('When button is clicked the app gets redirect', async () => {
    const previousPath = wrapper.vm.$route.path;
    const button = wrapper.find('[data-test-id="greetingsButton"]');
    await button.trigger('click');

    expect(wrapper.vm.$route.path).not.toEqual(previousPath);
  });

  it('The new route contains the name property', async () => {
    const button = wrapper.find('[data-test-id="greetingsButton"]');
    wrapper.setProps({
      name: 'Digitiamo',
    });
    await button.trigger('click');

    expect(wrapper.vm.$route.params.name).toEqual(wrapper.props().name);
  });
});
