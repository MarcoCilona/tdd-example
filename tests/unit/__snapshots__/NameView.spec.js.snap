// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewView tests Components does exists 1`] = `
<div>
  <p data-test-id="paragraphTest">Hello </p> <button data-test-id="greetingsButton">More greeting please!</button>
</div>
`;
